File: 0930-binary-subarrays-with-sum/0930-binary-subarrays-with-sum.py
class Solution:
    def numSubarraysWithSum(self, nums: List[int], goal: int) -> int:
        l1 = l2 = sum1 = sum2 = ans = 0
        for r, num in enumerate(nums):
            sum1 += num
            sum2 += num
            while sum1 > goal:
                sum1 -= nums[l1]
                l1 += 1
            while sum2 >= goal and l2 <= r:
                sum2 -= nums[l2]
                l2 += 1
            ans += l2 - l1
        return ans

File: 0930-binary-subarrays-with-sum/0930-binary-subarrays-with-sum.js
/**
 * @param {number[]} nums
 * @param {number} goal
 * @return {number}
 */
var numSubarraysWithSum = function(nums, goal) {
  return atMost(nums, goal) - atMost(nums, goal - 1);
};

const atMost = (nums, k) => {
  if (k < 0) return 0;
  
  let [l, r, ans] = [0, 0, 0];
  
  while (r < nums.length) {
    k -= nums[r];
    while (k < 0) k += nums[l++];
    
    ans += r - l + 1;
    r++;
  }
  
  return ans;
}